### PR TITLE
Bert Work in progress

### PR DESCRIPTION
**Pull request by benleetownsend**

### CODE REVIEW

File: finetune/base_models/bert/featurizer.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        )
+        output_state = {
+            'embed_weights': bert.get_embedding_table(),
+            'features': bert.get_pooled_output(),
+            'sequence_features': bert.get_sequence_output(),
+            'pool_idx': lengths
+        }
+        if config.num_layers_trained == 0:
+            print("STOPPING ALL THE GRADIENTS")
+            exit()
```
Comment (benleetownsend) [inline finetune/base_models/bert/featurizer.py]: Oops, remove the exit()

File: finetune/saver.py [hunk 2]
```diff
@@ -136,11 +136,14 @@ def init_fn(scaffold, session):
             self.var_val = []
             for var in all_vars:
                 for saved_var_name, saved_var in itertools.chain(variables_sv.items(), self.fallback.items()):
-                    if saved_var_name == var.name:
+                    if saved_var_name == var.name or saved_var_name == var.name[:-2]: # change this back and fix bert base :
```
Comment (madisonmay) [inline finetune/saver.py]: Would appreciate a little bit of context here

File: finetune/base_models/bert/featurizer.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
+import tensorflow as tf
+from finetune.base_models.bert.modelling import BertConfig, BertModel
```
Comment (madisonmay) [inline finetune/base_models/bert/featurizer.py]: *modeling

File: finetune/saver.py (hunk 2 above)
Comment (benleetownsend) [reply finetune/saver.py]: This needs removing. the first set of jl files I generated without `:0` on the end

File: finetune/nn/target_blocks.py [hunk 4]
```diff
@@ -290,11 +290,15 @@ def sequence_labeler(hidden, targets, n_targets, config, pad_id, multilabel=Fals
     """
     with tf.variable_scope('sequence-labeler', reuse=reuse):
         nx = config.n_embed
+
         def seq_lab_internal(hidden):
-            attn_fn = functools.partial(attn, scope="seq_label_attn", n_state=nx, n_head=config.seq_num_heads,
-                                            resid_pdrop=config.resid_p_drop, attn_pdrop=config.attn_p_drop,
-                                            train=train, scale=False, mask=False)
-            n = norm(attn_fn(hidden) + hidden, 'seq_label_residual')
+            if config.base_model.is_bidirectional:
```
Comment (lhz1029) [inline finetune/nn/target_blocks.py]: It might be worth modifying the docstring slightly to also apply to BERT. I'm specifically thinking of the line `The representations fed into this model are necessarily future masked because a language modelling loss is the original objective of the featurizer.`

File: finetune/saver.py [hunk 5]
```diff
@@ -141,6 +141,7 @@ def init_fn(scaffold, session):
                             saved_var = func(var.name, saved_var)
                         var.load(saved_var, session)
                         break
+
```
Comment (lhz1029) [inline finetune/saver.py]: Extra line

File: finetune/base_models/bert/bert_checkpoint_to_numpy.py [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+        ("multi_cased_L-12_H-768_A-12", "small_multi_cased"),
+        ("cased_L-12_H-768_A-12", "small_cased"),
+        ("cased_L-24_H-1024_A-16", "large_cased")
+
+    ]
+    for folder, out in models:
+        reader = tf.train.NewCheckpointReader(PATTERN.format(folder))
+        keys = list(reader.get_variable_to_shape_map().keys())
+        output_dict = {'model/featurizer/' + k + ":0": reader.get_tensor(k) for k in keys}
+        jl.dump(output_dict, OUTPUT_NAME.format(out))
```
[the remark this answers is not in the document]
Comment (benleetownsend) [reply finetune/base_models/bert/bert_checkpoint_to_numpy.py]: `reader.get_tensor()` a gets the tensor in numpy format from the checkpoint. This is unfortunately totally undocumented. https://www.tensorflow.org/api_docs/python/tf/train/NewCheckpointReader. Should probably just drop this file from the PR though. Or given an interface, wdyt?

File: tests/test_classifier.py [hunk 7]
```diff
@@ -27,6 +29,7 @@
 
 SST_FILENAME = "SST-binary.csv"
 
+SKIP_LM_TESTS = get_config().base_model.is_bidirectional
```
Comment (madisonmay) [inline tests/test_classifier.py]: I like this, feels super clean.

File: finetune/base_models/bert/bert_checkpoint_to_numpy.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (lhz1029) [reply finetune/base_models/bert/bert_checkpoint_to_numpy.py]: I think it would be useful as an interface, but leaving it as is in a script is not the worst thing etiher
